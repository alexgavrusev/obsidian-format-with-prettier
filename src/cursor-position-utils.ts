import { EditorPosition } from "obsidian";

export const editorPositionToCursorOffset = (
  position: EditorPosition,
  text: string,
): number => {
  const lines = text.split("\n");

  // if the cursor is on line i, on character j, the offset
  // is j plus sum of lengths (incl. the \n at the end) of lines 0..(i - 1)
  return lines
    .slice(0, position.line)
    .reduce((acc, line) => acc + line.length + 1, position.ch);
};

export const cursorOffsetToEditorPosition = (
  cursorOffset: number,
  text: string,
): EditorPosition => {
  const textUpToOffset = text.slice(0, cursorOffset);
  const lines = textUpToOffset.split("\n");

  return {
    line: lines.length - 1,
    ch: lines[lines.length - 1].length,
  };
};
