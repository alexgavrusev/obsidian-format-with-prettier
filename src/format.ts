import type { Options } from "prettier";

import * as prettier from "prettier/standalone";
import * as acornPlugin from "prettier/plugins/acorn";
import * as angularPlugin from "prettier/plugins/angular";
import * as babelPlugin from "prettier/plugins/babel";
import * as estreePlugin from "prettier/plugins/estree";
import * as flowPlugin from "prettier/plugins/flow";
import * as glimmerPlugin from "prettier/plugins/glimmer";
import * as graphqlPlugin from "prettier/plugins/graphql";
import * as htmlPlugin from "prettier/plugins/html";
import * as markdownPlugin from "prettier/plugins/markdown";
import * as postcssPlugin from "prettier/plugins/postcss";
import * as typescriptPlugin from "prettier/plugins/typescript";
import * as yamlPlugin from "prettier/plugins/yaml";

export type FormatArgs = {
  text: string;
  filepath: string;
  cursorOffset: number;
  prettierOptions: Options;
};

export const format = ({
  text,
  filepath,
  cursorOffset,
  prettierOptions,
}: FormatArgs) =>
  prettier.formatWithCursor(text, {
    filepath,
    cursorOffset,
    plugins: [
      acornPlugin,
      angularPlugin,
      babelPlugin,
      estreePlugin,
      flowPlugin,
      glimmerPlugin,
      graphqlPlugin,
      htmlPlugin,
      markdownPlugin,
      postcssPlugin,
      typescriptPlugin,
      yamlPlugin,
    ],
    ...prettierOptions,
  });
