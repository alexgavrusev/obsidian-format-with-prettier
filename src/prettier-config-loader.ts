import { App, EventRef, Notice, TFile, normalizePath } from "obsidian";

import type { Options } from "prettier";

const PRETTIER_CONFIG_PATH = normalizePath("prettierrc.json");

export class PrettierConfigLoader {
  private options: Options = {};

  constructor(
    private readonly app: App,
    private readonly registerEvent: (eventRef: EventRef) => void,
  ) {}

  private async readPrettierConfigFile() {
    const file = this.app.vault.getAbstractFileByPath(PRETTIER_CONFIG_PATH);

    if (!file || !(file instanceof TFile)) {
      return {};
    }

    const fileContents = await this.app.vault.read(file);
    const options = JSON.parse(fileContents) as Options;

    return options;
  }

  private async loadPrettierOptions() {
    try {
      const options = await this.readPrettierConfigFile();

      this.options = options;
    } catch {
      new Notice(
        `Failed to parse prettier options from ${PRETTIER_CONFIG_PATH}`,
      );

      this.options = {};
    }
  }

  async onload() {
    const handleConfigFileChange = (file: TFile) => {
      if (file.path === PRETTIER_CONFIG_PATH) {
        void this.loadPrettierOptions();
      }
    };

    this.app.workspace.onLayoutReady(() => {
      // this will be also triggered when the vault is 1st loaded
      this.registerEvent(this.app.vault.on("create", handleConfigFileChange));
      this.registerEvent(this.app.vault.on("delete", handleConfigFileChange));
      this.registerEvent(this.app.vault.on("modify", handleConfigFileChange));
    });

    await this.loadPrettierOptions();
  }

  getOptions() {
    return this.options;
  }
}
