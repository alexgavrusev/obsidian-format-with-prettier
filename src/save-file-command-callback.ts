import { App } from "obsidian";

export class SaveFileCommandCallback {
  private originalSaveCallback?: () => void;

  constructor(
    private readonly app: App,
    private readonly onFileSave: () => void,
  ) {}

  private getSaveCommandDefinition() {
    return this.app.commands?.commands?.["editor:save-file"];
  }

  onload() {
    const saveCommandDefinition = this.getSaveCommandDefinition();

    if (!saveCommandDefinition) {
      return;
    }

    // eslint-disable-next-line @typescript-eslint/unbound-method
    this.originalSaveCallback = saveCommandDefinition.callback;

    saveCommandDefinition.callback = () => {
      this.originalSaveCallback?.apply(saveCommandDefinition);

      this.onFileSave();
    };
  }

  onunload() {
    const saveCommandDefinition = this.getSaveCommandDefinition();

    if (!saveCommandDefinition) {
      return;
    }

    if (saveCommandDefinition.callback && this.originalSaveCallback) {
      saveCommandDefinition.callback = this.originalSaveCallback;
    }
  }
}
