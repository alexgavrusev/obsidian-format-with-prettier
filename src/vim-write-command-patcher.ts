import { App } from "obsidian";

/**
 * Patch the vim write command to trigger the
 * Obsidian file save command,
 *
 * @see https://forum.obsidian.md/t/vim-write-command-is-not-triggering-editor-save-file/40084
 */
export class VimWriteCommandPatcher {
  private originalCodeMirrorAdapterSaveCommand?: () => void;

  constructor(private readonly app: App) {}

  private getCodeMirrorCommands() {
    return window.CodeMirrorAdapter?.commands;
  }

  onload() {
    const codeMirrorAdapterCommands = this.getCodeMirrorCommands();

    if (!codeMirrorAdapterCommands) {
      return;
    }

    this.originalCodeMirrorAdapterSaveCommand =
      // eslint-disable-next-line @typescript-eslint/unbound-method
      codeMirrorAdapterCommands.save;

    codeMirrorAdapterCommands.save = () => {
      this.originalCodeMirrorAdapterSaveCommand?.apply(
        codeMirrorAdapterCommands,
      );

      this.app.commands?.executeCommandById("editor:save-file");
    };
  }

  onunload() {
    const codeMirrorAdapterCommands = this.getCodeMirrorCommands();

    if (!codeMirrorAdapterCommands?.save) {
      return;
    }

    if (this.originalCodeMirrorAdapterSaveCommand) {
      codeMirrorAdapterCommands.save =
        this.originalCodeMirrorAdapterSaveCommand;
    } else {
      // the original command was `undefined`, thus, the proper full cleanup would be deleting the key
      delete codeMirrorAdapterCommands.save;
    }
  }
}
