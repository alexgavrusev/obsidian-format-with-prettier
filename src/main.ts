import {
  App,
  Editor,
  Notice,
  Plugin,
  PluginSettingTab,
  Setting,
} from "obsidian";

import {
  cursorOffsetToEditorPosition,
  editorPositionToCursorOffset,
} from "./cursor-position-utils";
import { VimWriteCommandPatcher } from "./vim-write-command-patcher";
import { PrettierConfigLoader } from "./prettier-config-loader";
import { SaveFileCommandCallback } from "./save-file-command-callback";
import { format } from "./format";

interface PrettierPluginSettings {
  formatOnSave: boolean;
}

const DEFAULT_SETTINGS: PrettierPluginSettings = {
  formatOnSave: true,
};

export default class PrettierPlugin extends Plugin {
  settings: PrettierPluginSettings;

  private onFileSave = () => {
    if (!this.settings.formatOnSave) {
      return;
    }

    const editor = this.app.workspace.activeEditor?.editor;

    if (!editor) {
      return;
    }

    void this.formatFile(editor);
  };

  private readonly vimWriteCommandPatcher = new VimWriteCommandPatcher(
    this.app,
  );
  private readonly prettierConfigLoader = new PrettierConfigLoader(
    this.app,
    (ref) => {
      this.registerEvent(ref);
    },
  );
  private readonly saveFileCommandCallback = new SaveFileCommandCallback(
    this.app,
    this.onFileSave,
  );

  async onload() {
    await this.loadSettings();

    this.addSettingTab(new PrettierSettingTab(this.app, this));

    this.addCommands();

    await this.prettierConfigLoader.onload();
    this.vimWriteCommandPatcher.onload();
    this.saveFileCommandCallback.onload();
  }

  onunload() {
    this.vimWriteCommandPatcher.onunload();
    this.saveFileCommandCallback.onunload();
  }

  async loadSettings() {
    this.settings = Object.assign(
      {},
      DEFAULT_SETTINGS,
      (await this.loadData()) as PrettierPluginSettings | null,
    );
  }

  async saveSettings() {
    await this.saveData(this.settings);
  }

  private addCommands() {
    this.addCommand({
      id: "format-file",
      name: "Format current file",
      editorCheckCallback: (checking, editor, ctx) => {
        const file = ctx.file;

        if (!file) {
          return false;
        }

        if (checking) {
          return true;
        }

        void this.formatFile(editor);
      },
    });
  }

  private async formatFile(editor: Editor) {
    const file = this.app.workspace.getActiveFile();

    if (!file) {
      return;
    }

    const text = editor.getValue();

    try {
      const {
        formatted: formattedText,
        cursorOffset: formattedTextCursorOffset,
      } = await format({
        text,
        filepath: file.path,
        cursorOffset: editorPositionToCursorOffset(editor.getCursor(), text),
        prettierOptions: this.prettierConfigLoader.getOptions(),
      });

      editor.setValue(formattedText);
      editor.setCursor(
        cursorOffsetToEditorPosition(formattedTextCursorOffset, formattedText),
      );
    } catch (e) {
      new Notice(
        "Failed to format file, see the Developer console for more details",
      );
      console.error(e);
    }
  }
}

class PrettierSettingTab extends PluginSettingTab {
  plugin: PrettierPlugin;

  constructor(app: App, plugin: PrettierPlugin) {
    super(app, plugin);
    this.plugin = plugin;
  }

  display(): void {
    const { containerEl } = this;

    containerEl.empty();

    new Setting(containerEl).setName("Format on save").addToggle((toggle) =>
      toggle
        .setValue(this.plugin.settings.formatOnSave)
        .onChange(async (value) => {
          this.plugin.settings.formatOnSave = value;
          await this.plugin.saveSettings();
        }),
    );
  }
}
