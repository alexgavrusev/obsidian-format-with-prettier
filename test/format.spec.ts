import { describe, it, expect } from "vitest";
import { stripIndent } from "common-tags";

import { FormatArgs, format } from "../src/format";

describe("format", () => {
  describe("embedded languages", () => {
    const filepath = "note.md";
    const cursorOffset = 1;
    const prettierOptions = {};

    const getArgs = (text: string): FormatArgs => ({
      filepath,
      cursorOffset,
      prettierOptions,
      text,
    });

    const testFormat = (text: string) => format(getArgs(text));

    it("should format js", async () => {
      await expect(
        testFormat(
          stripIndent`
        # JS note

        ${"```javascript"}
          const foo = 'bar'



          const array = [1,    2,
            3];
        ${"```"}
        `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format jsx", async () => {
      await expect(
        testFormat(
          stripIndent`
        # JSX note

        ${"```jsx"}
          function MyButton() {
            return (
          <button>I'm a button


              </button>
            );
          }
        ${"```"}
        `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format ts", async () => {
      await expect(
        testFormat(
          stripIndent`
        # TS note

        ${"```typescript"}
          const foo:   string = 'bar'



          const array: Array<
          number > = [1,    2,
            3];
        ${"```"}
        `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format tsx", async () => {
      await expect(
        testFormat(
          stripIndent`
        # TSX note

        ${"```tsx"}
          function MyButton({title}:
                            { title: string }) {
            return (
          <button>{
            title
          }</button>
            );
          }
        ${"```"}
        `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format angular", async () => {
      await expect(
        testFormat(
          stripIndent`
        # Angular note

        ${"```typescript"}
          import {Component} from '@angular/core';
          
          @Component({
            selector: 'app-root',
            standalone: true,
            imports: [],
            template: \`
              <h1>



              {{


                title
              }}


              </h1>
            \`,
            styleUrls: ['./app.component.css'],
          })
          export class AppComponent {
            title = 'default';
          }
        ${"```"}
        `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format json", async () => {
      await expect(
        testFormat(
          stripIndent`
        # JSON note

        ${"```json"}
        {"foo": "bar", 
          "bar": true,

              "baz": 1
        }

        ${"```"}

          `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format jsonc", async () => {
      await expect(
        testFormat(
          stripIndent`
        # JSON note

        ${"```jsonc"}
        {"foo": "bar", 
          // a comment
          "bar": true,

              "baz": 1
        }

        ${"```"}

          `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format flow", async () => {
      await expect(
        testFormat(
          stripIndent`
        # Flow note

        ${"```js"}
          // @flow
          opaque type ID = string;
          
          function identity(x:      ID):ID {return x;
          }
          export type {ID};
        ${"```"}
          `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format handlebars", async () => {
      await expect(
        testFormat(
          stripIndent`
        # Handlebars note

        ${"```handlebars"}
          <ul class="people_list">{{#each people}}
                            <li>{{this}}</li>{{/each}}
              </ul>
        ${"```"}
          `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format graphql", async () => {
      await expect(
        testFormat(
          stripIndent`
        # GraphQL note

        ${"```graphql"}
          {
            project (

              name: "GraphQL"

            ) {
            tagline
            }
          }
        ${"```"}
          `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format html", async () => {
      await expect(
        testFormat(
          stripIndent`
        # HTML note

        ${"```html"}
          <h1>
          The Crushing Bore</h1><p>By Chris Mills</p>
          
          <h2>
                    Chapter 1: The dark night</h2>


          <p>It was a dark night. Somewhere, an owl hooted. The rain lashed down on theâ€¦
          </p>
        ${"```"}
          `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format css", async () => {
      await expect(
        testFormat(
          stripIndent`
        # CSS note

        ${"```css"}
          h1 {color: red;   font-size: 5em;
          }
          
          p {
          
          
          
            color: black;}
        ${"```"}
          `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format scss", async () => {
      await expect(
        testFormat(
          stripIndent`
        # SCSS note

        ${"```scss"}
          @mixin theme($theme:DarkGray  ){
            background: $theme;box-shadow: 0 0 1px rgba($theme, .25);
            color: #fff;
          }
          
          .info {@include theme;
          }
        ${"```"}
          `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format less", async () => {
      await expect(
        testFormat(
          stripIndent`
        # LESS note

        ${"```less"}
          #library() { .colors() {primary: green;
              secondary: blue;
            }
          }
          
          .button {
            color: #library.colors[primary ];border-color: #library.colors[secondary];
          }
        ${"```"}
          `,
        ),
      ).resolves.toMatchSnapshot();
    });

    it("should format yaml", async () => {
      await expect(
        testFormat(
          stripIndent`
        # YAML note

        ${"```yaml"}
          name: Martin D'vloper #key-value
          
          
          
          age:     26
          hobbies:
            - painting #first list item
            
            
            - playing_music #second list item
            - cooking #third list item
        ${"```"}
          `,
        ),
      ).resolves.toMatchSnapshot();
    });
  });
});
