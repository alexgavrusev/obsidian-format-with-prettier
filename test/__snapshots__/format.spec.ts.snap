// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`format > embedded languages > should format angular 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# Angular note

\`\`\`typescript
import { Component } from "@angular/core";

@Component({
  selector: "app-root",
  standalone: true,
  imports: [],
  template: \`
    <h1>
      {{ title }}
    </h1>
  \`,
  styleUrls: ["./app.component.css"],
})
export class AppComponent {
  title = "default";
}
\`\`\`
",
}
`;

exports[`format > embedded languages > should format css 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# CSS note

\`\`\`css
h1 {
  color: red;
  font-size: 5em;
}

p {
  color: black;
}
\`\`\`
",
}
`;

exports[`format > embedded languages > should format flow 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# Flow note

\`\`\`js
// @flow
opaque type ID = string;

function identity(x: ID): ID {
  return x;
}
export type { ID };
\`\`\`
",
}
`;

exports[`format > embedded languages > should format graphql 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# GraphQL note

\`\`\`graphql
{
  project(name: "GraphQL") {
    tagline
  }
}
\`\`\`
",
}
`;

exports[`format > embedded languages > should format handlebars 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# Handlebars note

\`\`\`handlebars
<ul class="people_list">{{#each people}}
    <li>{{this}}</li>{{/each}}
</ul>
\`\`\`
",
}
`;

exports[`format > embedded languages > should format html 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# HTML note

\`\`\`html
<h1>The Crushing Bore</h1>
<p>By Chris Mills</p>

<h2>Chapter 1: The dark night</h2>

<p>
  It was a dark night. Somewhere, an owl hooted. The rain lashed down on theâ€¦
</p>
\`\`\`
",
}
`;

exports[`format > embedded languages > should format js 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# JS note

\`\`\`javascript
const foo = "bar";

const array = [1, 2, 3];
\`\`\`
",
}
`;

exports[`format > embedded languages > should format json 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# JSON note

\`\`\`json
{
  "foo": "bar",
  "bar": true,

  "baz": 1
}
\`\`\`
",
}
`;

exports[`format > embedded languages > should format jsonc 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# JSON note

\`\`\`jsonc
{
  "foo": "bar",
  // a comment
  "bar": true,

  "baz": 1,
}
\`\`\`
",
}
`;

exports[`format > embedded languages > should format jsx 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# JSX note

\`\`\`jsx
function MyButton() {
  return <button>I'm a button</button>;
}
\`\`\`
",
}
`;

exports[`format > embedded languages > should format less 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# LESS note

\`\`\`less
#library() {
  .colors() {
    primary: green;
    secondary: blue;
  }
}

.button {
  color: #library.colors[primary];
  border-color: #library.colors[secondary];
}
\`\`\`
",
}
`;

exports[`format > embedded languages > should format scss 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# SCSS note

\`\`\`scss
@mixin theme($theme: DarkGray) {
  background: $theme;
  box-shadow: 0 0 1px rgba($theme, 0.25);
  color: #fff;
}

.info {
  @include theme;
}
\`\`\`
",
}
`;

exports[`format > embedded languages > should format ts 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# TS note

\`\`\`typescript
const foo: string = "bar";

const array: Array<number> = [1, 2, 3];
\`\`\`
",
}
`;

exports[`format > embedded languages > should format tsx 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# TSX note

\`\`\`tsx
function MyButton({ title }: { title: string }) {
  return <button>{title}</button>;
}
\`\`\`
",
}
`;

exports[`format > embedded languages > should format yaml 1`] = `
{
  "comments": [],
  "cursorOffset": 1,
  "formatted": "# YAML note

\`\`\`yaml
name: Martin D'vloper #key-value

age: 26
hobbies:
  - painting #first list item

  - playing_music #second list item
  - cooking #third list item
\`\`\`
",
}
`;
